sympy/assumptions/assume.py

```diff
@@ -159,7 +159,7 @@ def _hashable_content(self):
         return (self.name,)
 
     def __getnewargs__(self):
-        return (self.name,)
+        return (self.name, self.handlers)
```

Comment:
@oscarbenjamin, can you comment on this change, and whether `_hashable_content` should or should not include the handlers?

Reply:
The class seems broken to me:
https://github.com/sympy/sympy/blob/b1cc9eea67f71ca9223d8ac7a701305f4763bb4f/sympy/assumptions/assume.py#L152-L156
If this is a Basic subclass then all args should be passed up the chain to `Basic.__new__` unless this is an atom.

The `add_handler` method implies that the object is mutable so it should not subclass `Basic`.

Reply:
As I understand, `Predicate` is intended to be `Basic`. (See asmeurer's [comment](https://github.com/sympy/sympy/pull/20153#issuecomment-700523663)). I think we can make this `Atom`.

Reply:
Making it an atom doesn't change the fact that it's mutable.

My reading of @asmeurer's comment is that we should come up with a "clean design" but this is not it.

Reply:
Is it necessary for the handlers to be on the instance rather than the class? Why do they need to be pickled?

Reply:
As in #19953, it is better to implement function as an instance rather than a class, to treat it as a mathematical entity.

Reply:
I agree but only if there is a sensible way of using the function symbolically and if it is to be a Basic instance then it needs to be a proper `Basic` instance.